src/main/scala/shine/C/CodeGeneration/CodeGenerator.scala

```diff
@@ -163,6 +166,35 @@ class CodeGenerator(val decls: CodeGenerator.Declarations,
         case a: DataType => cmd(Lifting.liftDependentFunction[DataKind, CommType](fun.asInstanceOf[Phrase[DataKind `()->:` CommType]])(a), env)
       }
 
+      case DMatchI(x, inT, _, f, dPair) =>
+        exp(dPair, env, List(), input => {
+          // Create a fresh nat identifier, this will be bound to the "actual nat value"
+          val fstId = NatIdentifier(freshName("fstId"))
+          val sndT = DataType.substitute(fstId, x, inT)
+          val sndId = Identifier[ExpType](freshName("sndId"),  ExpType(sndT, `read`))
+          val (sndCType, initT) = typ(sndT) match {
+              case array:C.AST.ArrayType => (C.AST.PointerType(getBaseType(array)), (x:C.AST.Expr) => x)
+              case t => (t, (x:C.AST.Expr) => C.AST.UnaryExpr(C.AST.UnaryOperator.*, x))
+            }
+          // Read off the first uint32_t as a nat
+          C.AST.Block(immutable.Seq(
+            C.AST.DeclStmt(
+              C.AST.VarDecl(fstId.name, C.AST.Type.u32, Some(C.AST.ArraySubscript(C.AST.Cast(C.AST.PointerType(C.AST.Type.u32), input), C.AST.Literal("0"))))
+            ),
+            C.AST.DeclStmt(
+              C.AST.VarDecl(sndId.name, sndCType,
+                Some(initT(C.AST.Cast(sndCType, C.AST.BinaryExpr(input, C.AST.BinaryOperator.+, C.AST.Literal("4")))))
+            )),
+            cmd(f(fstId)(sndId), env.updatedIdentEnv((sndId, C.AST.DeclRef(sndId.name))))
+          ))
+        })
```

Comment:
would it be better/possible to use a C struct instead of casts? (potentially re-using regular tuple codegen)

Reply:
I don't think so. A nat-dependent pair with an array element is pretty much a dynamic-sized array whose length is encoded as a payload in the first element.
For now, the strategy is I am going for it represent the type `(Nat ** (n => T(n)))` (dependent pair where the first element is a nat and the second element some data-type T which may depend on n), is to represent it as a single `uint8_ t* buffer`, from which the first element is projected out at offset 0 (by doing `uint32_t first = ((uint32_t*)buffer)[0]`, and then use that identifier to know the size of the second element.

A struct doesn't work, because we'd need conceptually something like
``` struct {  T1 n; T2 something[E(n)] }```. where E is some arithmetic expression.

Let me know if this makes sense, or I have misunderstood you along the way

(PS - originally, I wanted to represent them with structs of pointers to data. This would allow an easier access, but that would mean that a dependent  pair is a struct + two buffers, and that makes GPU-CPU interaction hard, and also  adds further complications when it comes to allocating these things. So ultimately, I believe the "one (materialised) pair is one buffer" approach works best)

Reply:
Would the struct not look more like `struct { T1 n; T2* array; }`, where the size of the array is erased in C? Although this is only for arrays, maybe it does not work for the general case. You could also have `struct { T1 n, uint8_t* buffer }`, although I am not sure if that would bring much benefit. Maybe you could use a combination of tuples and casts in imperative DPIA to express the dependent pair logic without changing the lowest level codegen.

Reply:
With that struct approach, I lose the property that 1 dependent pair = 1 buffer, no? The second element would be a pointer, not an actually buffer itself. That is quite a radical departure from our model, where 1 "concrete" value is actually 1 "contiguous" buffer.

About elimination of dependent patterns before codegen, I see the value of it, but I am not sure how to actually do it and preserving  the properties I want.

That being said, it might be worth it to perhaps  push some of the complexity out of codegen and into the imperative language, I guess. This I think may be done. 

About turning things into pairs, I think this won't be possible as long as records (materialised pairs) are represented as C structs. I believe our code generator system already falters if you try to generate a record of arrays. This I think has been fine so far as we never wanted to do such a thing - but with dependent pairs, that's kind of the whole point

Reply:
Sounds reasonable. I guess it would be easier to start by pushing things like split/join out of the monolithic codegen rather than this.

Reply:
Actually about that, I wonder how much do you think should we push out? I can think it's  at least two different "levels" of work here: one about turning `split/join` into multi-dimensional indexing, and a further level of index flattening.  Would you ideally like both to be done as transforms over imperative `DPIA`?

Reply:
It would be about turning `split/join` to multi-dimensional indexing, breaking it down to `idx`. The flattening can remain in the `DPIA -> C` part.

Reply:
Makes sense